### ISSUE TITLE
NSIS target

### ISSUE THREAD

**Issue by farfromrefug** [electron-userland/electron-builder#472]
May i ask why you dropped NSIS completely?

Dont get me wrong squirrel is nice but it can't do it all. And most of my clients don't really like/understand those installer which give them no choice (installation directory, ...).
Also squirrel does not seem to support installation of 3rdarpties (like boujour SDK, or any other needed drivers)

I could go and not use electron-builder at all but that's not what i want. 
I could fork electron builder and bring it back from where you dropped it, but again not really good for the community.

Could we talk about it? about bringing it back?

Thanks

**Comment by englishextra** [comment]
Yeah that was frustrating

**Comment by develar** [comment]
> Also squirrel does not seem to support installation of 3rdarpties

Your app can install anything after run (on first launch / on squirrel event).

> Could we talk about it? about bringing it back?

I cannot say that I am happy to use / maintain Squirrel.Windows.

And I understand that it is not developer responsibility to teach users. Windows is Windows.

Personally, as OS X user, I don't like installers. `pkg` is very rare on OS X (and often indicator that this developer/company is bad / don't think about users"). And I think Squirrel.Windows should be preferred way to distribute, and don't forget about auto-update.

In short: you can expect NSIS and 7z sfx targets next week.

**Comment by englishextra** [comment]
electron-builder came on the scene as NSIS dependancy, So lets leave it. The rest like squirrel dependant electron-builder should be renamed into new project. Its like your offered lemonade and brought milk coctail.

**Comment by farfromrefug** [comment]
@develar First thanks a lot for taking the time to answer. And i am really happy to to see NSIS coming back.
As an OSX user i really don't like installers either, but Windows and OSX are totally different in the way you integrate 3rdParties. I thought about running the installer on first run but i don't  really like it. It means i either have to package the 3rdparty installer in my app package or get it from the web. Not really fond of it. I like the installer becasue that's the way to do it on windows... Just like deb is the way to do it on linux.
Plus install location is kind of a big issue on windows. Especially in companies where there are permission issues. And there is also "permission actions". Which is mostly why you have pkg on OSX.
I think that's why big projects decide to stay with NSIS or else like VSCode for example.

I see squirrel as the best solution for portable install with installer like it is on OSX.
By the way about the update system. You can use sparkle or winsparkle, which i think i am gonna go for.

I have already looked (a few years ago, just coming back to desktop dev) at other solutions. Nothing much better right now i think.

Thanks again @develar

**Comment by develar** [comment]
> By the way about the update system. You can use sparkle or winsparkle, which i think i am gonna go for.

NSIS is enough — just download and run installer `silent` – https://github.com/codeskyblue/electron-inno-auto-update (this example for inno setup, but the same for NSIS). And you can use standard electron auto-update API — no vendor lock.

**Comment by develar** [comment]
NSIS seems to outperform Squirrel.Windows:

* 50% less installer size (~ 30 MB vs 60 MB)
* 50% and more (not yet tested) less delta update if build on non-Windows platform. (Squirrel.Windows delta is efficient only if build on Windows).
* easy for user — declarative support of desktop shortcuts, file associations.
* native (no Wine!!!) installer builder.

It will require work on our side, though, but it seems it is better to invest into NSIS, than fix "broken" (sorry, sorry) Squirrel.Windows.

**Comment by farfromrefug** [comment]
@develar That answers my question https://github.com/Microsoft/vscode/issues/7227 !

**Comment by englishextra** [comment]
@farfromrefug  - confirm that nsis and so does inno have a small installer size - my app files take somewhat of 5 mb and the final exe of nsis and inno is about 33mb
when compiled to msi with adancedinstaller - i ts around 50mb
@develar good news of yours about nsis revival

**Comment by JimiC** [comment]
:+1: for bringing back NSIS. Although @paulcbetts work on Squirrel is awesome, it just doesn't cut it for me. Don't get me wrong, maybe Squirrel will become "must-have" in the future, but for now it just doesn't suffice.

**Comment by develar** [comment]
@JimiC Yes, NSIS outperforms Squirrel.Windows in all aspects — it requires the same work to make it excellent as fixing Squirrel.Windows (but result will be much more stable and robust, instead of maintaining our own fork of Squirrel.Windows). Maybe we even make NSIS by default.

**Comment by paulcbetts** [comment]
Electron will always be using Squirrel as its installer for the foreseeable future. If electron-builder wants to integrate NSIS or another installation framework instead they're of course entirely in their right to do so

**Comment by develar** [comment]
@paulcbetts Current plan just add additional target.

> Electron will always be using Squirrel as its installer for the foreseeable future

If I understand correctly, it is only bundled auto-update API implementation, right?

**Comment by paulcbetts** [comment]
> If I understand correctly, it is only bundled auto-update API implementation, right?

Correct

**Comment by englishextra** [comment]
I dont need VS and NUGET in my work. I need Phonegap to build webapp in APK / XAP. And I need electron-packager to build an app for Windows / Linux, and I need electron-builder to distribute those Windows apps. If electron-builder fell (sorry) to Squirrel.Windows strory, I would have to use NSIS/INNO to build installers. That's what I'm doing now. Inno Setup seems better than NSIS for some reasons. In any case electron-builder choses Squirrel - it loses a piece of pie.

**Comment by develar** [comment]
> In any case electron-builder choses Squirrel - it loses a piece of pie.

We don't choose anything. We are orchestration tool. I like Squirrel.Windows as user and for now it will be default windows target. The same as in case of Linux — by default deb, but you can build rpm, packman and other 5+ formats if you want.

> I dont need VS and NUGET in my work.

Squirrel.Windows doesn't require you to use it.  .net is not a problem in the modern Windows versions.

> Inno Setup seems better than NSIS for some reasons.

Please describe such reasons.

**Comment by englishextra** [comment]
Oh yeah? Did I misunderstand something - https://github.com/Squirrel/Squirrel.Windows
Building Squirrel
```
git clone https://github.com/squirrel/squirrel.windows
cd squirrel.windows
git submodule update --init --recursive       ## THIS IS THE PART YOU PROBABLY FORGOT
.\.NuGet\NuGet.exe restore
msbuild /p:Configuration=Release
```

**Comment by JimiC** [comment]
My .02 cents here. From my POV if I had build `electron-builder` I would had provided choices when it comes to windows installer builders, just like it does now with Linux. With a simple 'win' specific option (i.e packager ?) with available values either 'squirrel', 'nsis' or 'inno'. But I'm not the builder.

**Comment by mzmousa** [comment]
@englishextra you only need to do that if you're trying to run Squirrel by itself. When using `electron-builder`, it handles everything squirrel related for you. You don't need to get VS or Nuget in order for `electron-builder` to run successfully. That's why it comes with dependencies.

**Comment by JimiC** [comment]
@englishextra I personally tried to build `Squirrel.Windows` as I'm mainly a .NET developer and VS is my choice of IDE. Got stuck a little bit figuring out that I had to add the ILRepack nuget package (not inluded in the project), Windows SDK and C++ tools, but after that I had no issues. But not everyone knows its ways around Windows Software development, so I give you that.

**Comment by englishextra** [comment]
@mzuma Thanks. I  updated via npm electron-builder and launched the build process as usual, and it failed. That's why I started to install Squirrel.Windows. For me it's obvious - that such unstable situation has driven me simply writing install scripts by myself. This consumes much more time.
@JimiC Yes - I was asked to add nuget too, and actually I'd like to have control what resides on my machine - in case of NSIS/INNO - everything is clear - I use them on daily basis.

**Comment by develar** [comment]
@JimiC offtopic: electron-builder fork of Squirrel.Windows builds on CI server — see https://github.com/develar/Squirrel.Windows/blob/build/appveyor.yml To avoid any issues on development machine.

> I updated via npm electron-builder and launched the build process as usual, and it failed. That's why I started to install Squirrel.Windows.

In this case just report issue here before digging.

**Comment by englishextra** [comment]
@develar Thanks for reminding the procedure.

**Comment by mzmousa** [comment]
@englishextra Sorry, I meant to edit my comment but deleted it (git noob). What failed in your build process? I know this is probably not the place to ask about that but myself and other devs on my team have been using `electron-builder` and its been working magic. Squirrel.Windows is indeed a pain to deal with, but thankfully @develar dealt with it. I don't see how it would fail or give you a hard time unless you had a weird configuration, but [this blog](http://electron.rocks/electron-builder-explained/) by @Vj3k0 is excellent, and should smoouth out the process.

**Comment by englishextra** [comment]
@mzmousa The batch file with

````
electron-builder ./.dist/englishextra-app-win32-ia32 --platform=win --out=./.dist/ --config=builder.json
:: electron-builder ./.dist/englishextra-app-win32-x64 --platform=win --out=./.dist/ --config=builder.json

````
when lauched normally worked fine. A couple of days ago it stopped working - the cmd starts and closes. While electron-packager continues ok to build packages. You know I see no possibility to figure out what could happen. Then I updates electron-builder about 3 days ago - with no success again. Clients wait - I switched to InnoSetup and the company. You see I have no tools to debug where electron-builder gets stuck - since cmd simply goes down throwing no error.

**Comment by develar** [comment]
> You see I have no tools to debug where electron-builder gets stuck - since cmd simply goes down throwing no error.

Just _report_ issue. No report — no help or fix. _All_ tools has debug output if set `DEBUG` env.

**Comment by englishextra** [comment]
@develar Thank you very much for your time. Everything will work out just fine. This way or another.

**Comment by englishextra** [comment]
We could do it that way - I will install electron-builder and strat building again. And report to you the result shortly?

**Comment by JimiC** [comment]
@englishextra I believe `iconUrl` is mandatory for `win.build`. In my case I have only that the the options and win installer builds just fine.

**Comment by englishextra** [comment]
@JimiC More specific? About iconURL

OK 4.2.6 - No success. Please advise on debugging

**Comment by JimiC** [comment]
@englishextra https://github.com/electron-userland/electron-builder/wiki/Options#buildwin

**Comment by englishextra** [comment]
Well I did this 

```
	"win": {
		"title": "englishextra-app",
		"version" : "0.1.0",
		"publisher": "englishextra.github.io",
		"nsiTemplate": "installer.nsi.tpl",
		"iconUrl": "https://github.com/englishextra/englishextra.github.io/master/favicon.ico?raw=true"
	}
```
No success

BTW - I recall somebody was offering some debug flags

**Comment by JimiC** [comment]
@englishextra remove everything but the `iconUrl` and try again. By the way are you building on Windows or another env?

**Comment by mzmousa** [comment]
@englishextra what's `version` doing in there? Take it out.
If you meant to specify the electron version, do it in `build`, not in `win`, and do so like this:
```
"win": {
    "electronVersion": "1.2.1"
}
```

**Comment by englishextra** [comment]
Well I ended adding blob/

	"win": {
		"title": "englishextra-app",
		"version" : "0.1.0",
		"publisher": "englishextra.github.io",
		"nsiTemplate": "installer.nsi.tpl",
		"iconUrl": "https://github.com/englishextra/englishextra.github.io/blob/master/favicon.ico?raw=true"
	}

And also tried removeing this iconUrl key - no result.  Thanks anyway.

**Comment by englishextra** [comment]
D:\server\sources\englishextra-app\.electron\.dir.bat
D:\server\sources\englishextra-app\.electron\.dir.txt
D:\server\sources\englishextra-app\.electron\electron-v0.36.5-darwin-x64.zip
D:\server\sources\englishextra-app\.electron\electron-v0.36.5-linux-ia32.zip
D:\server\sources\englishextra-app\.electron\electron-v0.36.5-linux-x64.zip
D:\server\sources\englishextra-app\.electron\electron-v0.36.5-win32-ia32.zip
D:\server\sources\englishextra-app\.electron\electron-v0.36.5-win32-x64.zip

**Comment by JimiC** [comment]
@englishextra Didn't said to remove `iconUrl` but remove everything else BUT the `iconUrl`. Also read my comment again (about the ICO size).

**Comment by englishextra** [comment]
@JimiC I got you wrong - and I didn't referenced you when said I tried to remove icon key. 

Gents, Please point me to a sample build.json in your wiki and I will reproduce it tommorrow morning. You saw my build,json and I didn't hear a word about exact error except iconUrl. Can I have a look at sample json including iconUrl. Thaks to everyone.

**Comment by develar** [comment]
@englishextra I can speak russian. Feel free to contact me. And please let's stop off topic here ;)

**Comment by englishextra** [comment]
@develar Sure, you can— others cannot. 

The last thing is that Windows section here https://github.com/electron-userland/electron-builder/wiki/Multi-Platform-Build is 

> Not documented yet.

So I just switched to Watching mode. Thanks, no need to bother.

**Comment by mzmousa** [comment]
@englishextra 
> I can speak russian. Feel free to contact me. 
He meant that if you're stuck somewhere, and there's a language barrier for you _specifically_, he can help you in russian if you do need it. He was just trying to be nice.

**Comment by englishextra** [comment]
@mzmousa 
a) link to debugging methods described clearly
b) list of Windows dependancies
c) NSIS/INNO Setup support confirmation
This is what needed.

**Comment by kunkinkan** [comment]
As @JimiC said: 

` With a simple 'win' specific option (i.e packager ?) with available values either 'squirrel', 'nsis' or 'inno' (default could be 'squirrel').
`

is the solution to this issue. But squirrel should remain the default. It outperforms NSIS in ease-of-installation (nothing to click...) and that's what matters most.

**Comment by develar** [comment]
@kunkinkan Our Nsis target with default options will produce one-click installer as Squirrel does. But size will be 50% less, installer will have true progress bar, shortcuts and file associations in the declarative cross-platform way (so, installation will be a little bit faster). And will be ability to pack ia32 and x64 editions in one installer.

But of course nsis will be not default target for now.

**Comment by develar** [comment]
NSIS is amazing. Yes, it is far from ideal, but our NSIS one-click installer:
* smaller
* faster
* no more squirrel events in your code, no more platform-specific code (file associations and other stuff)
* true process bar #374 
* app user model id #362

Instead of fixing Squirrel.Windows (net code), you can easily modify text file and do what your want. It is amazing. It is implemented and will be pushed soon. 

@farfromrefug Thanks! I will not waste my time to fix Squirrel.Windows anymore. NSIS will be default target once it will be ready for production use.

It is one-click per-user installer as Squirrel.Windows, but options 

```
export interface NsisOptions {
  /*
  Mark "all users" (per-machine) as default. Not recommended. Defaults to `false`.
   */
  readonly perMachine?: boolean | null

  /*
   Allow requesting for elevation. If false, user will have to restart installer with elevated permissions. Defaults to `true`.
   */
  readonly allowElevation?: boolean | null

  readonly guid?: string | null

  /*
  One-click installation. Defaults to `true`.
   */
  readonly oneClick?: boolean | null
}
```

allows you to configure it or allow user to choose.

Source code is easy to customise (i.e. you don't need to copy the full script just to modify several lines), so, you do what you want (but keep in mind — it is better for all to open PR and add functionality using flags).

**Comment by kunkinkan** [comment]
Does this install in the same location as with squirrel.windows?

**Comment by develar** [comment]
@kunkinkan No exactly, because we must install not to `LocalAppData/AppName`, but to `LocalAppData/Programs/AppName` according to Win7+ standard (Win7+ has a per-user programfiles known folder). You can set `nsis.perMachine `  to `false` to change default.

**Comment by kunkinkan** [comment]
When will NSIS be available? I'll be releasing an app pretty soon, don't want to have 2 install locations. (unless there's some kind of squirrel removal done by NSIS ;-))

How will updates work without squirrel events?

**Comment by JimiC** [comment]
> The autoUpdater module provides an interface for the Squirrel framework.

http://electron.atom.io/docs/api/auto-updater/

**Comment by develar** [comment]
> electron.autoUpdater

electron.autoUpdater API will be supported in the same way as Squirrel.Windows and bundled (i.e. no external dependency in your app). In your app you will change the only line — URL of updates.

**Comment by develar** [comment]
> When will NSIS be available?

As we are open source project, it depends on you ;) I think #495 will be landed today. But autoupdate will be implemented later.

> unless there's some kind of squirrel removal done by NSIS ;-)

Because it is NSIS, it is very easy to do anything you want. Probably it will be implemented as well (PR is welcome).

**Comment by farfromrefug** [comment]
@develar great work on your part. Thanks a lot!
